### ISSUE TITLE
Is the libzerocash Merkle tree vulnerable to hash-for-data substitution?

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#404]
Amazon's implementation of Merkle trees is vulnerable to an attack:

https://mailarchive.ietf.org/arch/msg/openpgp/aQrQJe-iGvhQj8ziBo5GW1OrbO4

```
In the simple case, take any long input that's a power of two blocks
long.  Calculate its Amazon-style hash tree root value.  While
calculating it, remember the top two non-root internal node hash
values.  Concatenate them and compute the Amazon-style hash tree root
for *that*.  You'll get exactly the same hash tree root.
```

Is libzerocash's implementation similarly vulnerable?

**Comment by daira** [comment]
We believe there is no vulnerability for a fixed-depth tree. This would need to be revisited if we implemented #649, but that is "not in 1.0".

**Comment by daira** [comment]
See also #501.
